### ISSUE TITLE
unobstrusive function call not working

### ISSUE THREAD

**Issue by jbonlinea** [jonnyreeves/jquery-Mustache#40]
Hi,
I found this post of yours, thank's , very usefull for mustache beginers !
[ ](http://jonnyreeves.co.uk/2012/using-external-templates-with-mustachejs-and-jquery/ )

One question still :
Pre : my #tpl-greeting contain obstrusive and unobstrusive calls to javascript functions

- if I render my template by placing the renderer - `$get('greetings.html', function ... });` - within a script tag within my html main page, everything works fine,
- but if I place the renderer in an external .js file that I integrate to my main html file right at the same place than previously, the renderer do render, but only my functions obstrusivelly called are indeed called, the ones unobstrusively called are not called anymore (while the later still works when called obstrusively) !!!

Any clues on that ?

thank's in advance 

ps : I know it's not directly related to your plug-in but still is an issue for the implementation of the whole

**Comment by jonnyreeves** [comment]
Could you upload an example that reproduces the issue please?

**Comment by jbonlinea** [comment]
Hello

Thank's for your interest,let me try !

It's not exactelly in line with what precede because of further test and I took the whole out of the main project file containing many more things.

**test.html** 
everything is included in `test.html` 
- render ok
- myfunction ok

**test1.html**
the template is included in `test1.html`
the .js part is in `render_loc-funct.js`
- render ok
- myfunction ok


**test2.html** 
the template is in `tpl.html`
the .js part is in included in `test2.html`
- render ok, more precisely `myDiv` is populated with then content of  `tpl.html > #tpl-html`
- `myFunction` do not work anymore, whether called obtrusively or unobtrusively (but it's designed to be called unobtrusively so the former is no surprise)
- I created a `myFunction2` that I call obstrusivelly in the `last button` and it works (see the console).

This I don't understand.

---

Then,
As mentionned in my previous post, this afternoon, on my main project file, I did manage to render the external template `tpl.html` and call `myFunction` unobtrusively, but 
- only when the renderer is included in `main.html` not in my `main.js` file;
- I can't even reproduce this in `test2.html` ?! 

With further test on my main project file, I realized that :
- if my renderer is in my `main.js` file, wether **at the begining** or the end, mustache rendering is done, but I can't call `myFunction` unobstrusively
- if my renderer is in `main.html` **and before** `main.js`, everything is fine
- if my renderer is in `renderer.js` **and included before** `main.js`, everything is also fine

This I don't  really understand neither, even if I assume it is related to asynchronism 


Finally, this openend the door to a **test3.html** 
This test has four files : `test3.html`, `tpl.html`, `renderer.js`, `myfuntion.js`.
It will work **if and only if** `renderer.js` **is included before** `myfunction.js`
As you can see, `rendered.js` and `myfunction.js` **contain exactly the same code** than the two script included in `test2.html` **where it was not working**. 
Again, if I copy the content of `myfunct.js` at the botom of `renderer.js`, mustache rendering is done but I can't call `myFunction`... 

That's ... a day of crowling, and I don't know what has to be learned from it beyond that asynchronicity has an semi-eratic behaviour quite unpredictable with framework/plug-in

But if you have a better explaination, I'm more than willing to hear it 

Regards 
[Archive.zip](https://github.com/jonnyreeves/jquery-Mustache/files/664325/Archive.zip)

**Comment by jbonlinea** [comment]
ps : 
I've also been jungling with this approach [here](https://www.sitepoint.com/creating-html-templates-with-mustachejs/)
and now realize that when I keep the `document.ready` event listener  `myFunction` can't be called unobstrusively while it may whithou this event listener, so again **asynchronicity seems to be the master word.**

However as I'm a beginer coder and only discover jQuejy (better with javascript) I do not see the pro and con of your two approaches : 
- `$.**get**('tpl.html', function(templates) {}`
vs
- `$("#templates").**load**("tpl.html #tplhtmldiv",function(){}`

even if I understant that your technique do not `need a hidden buffer div` to pre-load the template 

Thank's again !

**Comment by jbonlinea** [comment]
UPDATE : 
test3.html worked only in my web editing tool, not in firefox, nor chrome... 

So indeed, in the actual state of the code, can't unobstrusively call a function by clicking on an element in an external template ! ?

**Comment by jonnyreeves** [comment]
Hi @jbonlinea, sorry for the delay.  Please find the attached zip file which provides a basic implementation of scenario 3 (external templates with the `onclick` handlers applied using late-binding).

The trick you were missing was to wait for the `document` to fire its `ready` event before attempting to bind the click handlers.

[jquery-Mustache-issue-40.zip](https://github.com/jonnyreeves/jquery-Mustache/files/673824/jquery-Mustache-issue-40.zip)

**Comment by jbonlinea** [comment]
Hi @jonnyreeves,

Happy new year and thank you very much for your involvement !

I just did few test, and it seems that your code only takes us half way.

In firefox, after the first page load, the function wasn't called using the unobstrusive approach, while after reloading the page, it did worked...

So we do still have a script loading order requirement issue and as explained [here](http://stackoverflow.com/questions/950087/how-do-i-include-a-javascript-file-in-another-javascript-file) 
> you won't be able to use your newly loaded code the next line after you asked it to be loaded

In the meantine, I used lab.js. and expressed an execution order dependency by inserting `.wait()` 
`<script src="https://cdnjs.cloudflare.com/ajax/libs/labjs/2.0.3/LAB.min.js"></script><script type="text/javascript">$LAB .script("https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.1/jquery.min.js").wait() .script("https://cdnjs.cloudflare.com/ajax/libs/mustache.js/2.3.0/mustache.min.js").wait() </script> <script type="text/javascript"> $LAB .script("firstScript.js").wait() .script("myMainJs.js").wait(function(){initMyPage(); }); </script>`

It did solved all off my other, but similar, issue related to script loading and requierment but this one (unobstrusively call a function from an external template).
I did a quick try this morning, with your code, and some other tries, but still no chance...

I belive we may find a solution by spliting the js in different files and load them as needed, but it seems quite a lot of fuss ?! 

cheers
